pkg/cmd/server/origin/run_components.go

```diff
@@ -376,7 +376,7 @@ func (c *MasterConfig) RunDeploymentImageChangeTriggerController() {
 // RunSDNController runs openshift-sdn if the said network plugin is provided
 func (c *MasterConfig) RunSDNController() {
 	oClient, kClient := c.SDNControllerClients()
-	controller, err := sdnfactory.NewMasterPlugin(c.Options.NetworkConfig.NetworkPluginName, oClient, kClient)
+	controller, err := sdnfactory.NewMasterPlugin(c.Options.NetworkConfig.NetworkPluginName, oClient, kClient, c.EtcdHelper)
```

Comment:
- No change to clients: This is undesirable (1) Old client calls are
   synchronous and to ensure backward compatibility, server needs to poll to
   get assigned vnid for namespace (we want this to happen on the client side).


  So current clients don't retry if there is no vnid?  That's broken and we
  should fix that before 3.3 by releasing a fix in 3.2.1.  That's a much
  smaller risk migration issue.

=> Old clients do *not* need to retry because it only depends on NetNamespace etcd updates which is a sync op. New clients need to retry because asking for vnid is a request and this request is handled by vnid controller asynchronously (picks requests from single queue to ensure correctness).


   - (2) This introduces unnecessary indirection and complicates feature
   addition. Upcoming features like isolate-projects (undo join-projects) and
   ingress policy(upstream) stores info as annotations on namespace. To
   support this functionality via pod-network cli cmd in this case will need
   to pass annotation to NetNamespace first and then propagate to Namespace.


  No, you would just call the net namespace API, and the registry would
  handle it.  We've dealt with scenarios like this in the past and this is
  the most manageable approach we've found.

=> Yes, we could make it work but definitely needs more work for adding/maintaining features.

   - I guess migrator is expected to run after upgrade of all nodes and
   masters. Once migrator is called, it will perform copy to namespace and
   deletes object but any create calls after that will still create etcd based
   NetNamespace objects? How long we will maintain this?

  Until the admin sets a config flag, or we switch the controller to allocate
  only via namespace.  Who calls create today?

=> Currently, create is called by SDN master whenever it detects a new namespace. In the new code, vnid controller watches for namespaces and if it detects any namespace without a vnid, it allocates a new one (no explicit create call involved).

=> Also realized migrator framework may not work in the current approach. If master is upgraded and if we don't do migration in place, then vnid controller will assign new vnids to  existing namespaces as those will not have vnid annotations. This will break shared/global project networks.


  If we don't want controller to have direct access to etcd then we could do:

   - By default, make NetNamespace act as virutal resource backed by
   namespace. Old client works without any issues.

  Only after a successful copy of the data from net namespaces to namespaces,
  which would be incomplete.  Old client only works if the data is in place,
  so this assumes a pre upgrade copy for existing namespaces.

=> As mentioned before, pre upgrade of existing namespaces is needed.

   - New nodes watches for both NetNamespaces and Namespaces, so new nodes
   can work with old masters.

  Nodes can't really watch namespaces - this is a scalability problem (nodes
  should not be holding in memory all namespaces).  You have to fetch
  namespaces on demand and cache if necessary.  I'd actually prefer from a
  security perspective if nodes didn't have access to namespaces, because
  there is no guarantee namespace info isn't sensitive.  Nodes need to be
  prevented from seeing that info - that's why NetNamespaces should continue
  to exist as a virtual resource even after we've migrated away the etcd data.

=> With the new code, nodes are not storing all namespaces in memory, they are reacting to vnid annotation changes on namespace. Upstream network policy is storing some stuff as annotation(e.g. net.alpha.kubernetes.io/network-isolation) along with NetworkPolicy objects, we may need to watch this resource any way. NetNamespace has single piece of info that is VNID (supposed to be internal) and our plan is to retire this object to make it inline with upstream approach of dealing with Namespaces and NetworkPolicy objects for networking. Backward compatibility is needed for 'oadm pod-network', NetNamespace is internal stuff (users should not care about this resource).

   - New master will migrate NetNamespaces (client using custom flag)
   [Migration order: nodes and then masters] This will be much cleaner for
   future changes. No down time during migration, maintains backward
   compatibility and don't need to deal with etcd backed NetNamespaces after
   migration (no dual sets of values from namespaces and netnamespaces).

  How is the master going to migrate that without access to etcd?  It would
  be better if update on the NetNamespaces ensured that the data was written
  to the namespace, or if we used that flag on the masters.

=> Master can do migration using NetNamespace client API. API will access etcd backed netnamespace if custom flag/annotation is passed (used only for migration) otherwise it will act as pure virtual resource backed by namespace (supports old clients)

  Ultimately, all schema migrations are four phases:

  1. Write both old and new values (in this case etcd object and onto
namespace)
  2. When all code is upgraded, finish upgrading the old data (migrator via
  the API)
  3. Flip a switch to only read and write new data (via config or etcd).
  4. Delete old data

  Nodes shouldn't have to change in this migration (except maybe to retry as
  noted).  Keep the client API consistent.

  Use the internal flag or etcd value to alter the behavior of the registry
  to make the API alter.  Write both paths (forcing client to retry).  Run
  migrator once all apiservers are upgraded.  Flip config flag once
  everything is migrated.  Once all new API servers are running with the new
  flag, delete old data from etcd.

=> I like the general migration framework but in this particular case I don't think this is the right approach. Making NetNamespace to fetch dual sets of values from namespaces and netnamespaces will not enable us to retire this resource(one of the primary intent). As mentioned before, this resource is supposed to be internal/not-user-facing.

@smarterclayton @liggitt @knobunc @dcbw @danwinship @rajatchopra